### PR TITLE
mgr/stats: missing clients in perf stats command output.

### PR DESCRIPTION
**Pull request by neesingh-rh**
perf stats doesn't get the client info
-when new filesystems created or filesystems created on failing other filesystem
after running the perf stats command once with existing filesystems.
-if there are 2 filesystems mounted, failing the second filesystem's
rank0 mds shows missing/stale client metrics.

Fixes: https://tracker.ceph.com/issues/56483
Signed-off-by: Neeraj Pratap Singh <neesingh@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by vshankar** [comment]
> perf stats doesn't get the client info w.r.t new filesystems
> created or filesystems created on failing other filesystem
> after running the perf stats command once with existing filesystems.

Add a test case for this.

**Review by vshankar — CHANGES_REQUESTED** [review]
@neesingh-rh

File: qa/tasks/cephfs/test_mds_metrics.py [hunk 1]
```diff
@@ -511,3 +511,60 @@ def test_client_metrics_and_metadata(self):
...
+        self.mount_a.create_files()
+
+        # creating another filesystem
+        fs_b = self._setup_fs(fs_name="fs2")
+
+        # Mount a client on fs_b
+        self.mount_b.mount_wait(cephfs_name=fs_b.name)
+        self.mount_b.write_n_mb("test.bin", 1)
+        self.mount_b.path_to_ino("test.bin")
+        self.mount_b.create_files()
```
Comment (joscollin) [inline qa/tasks/cephfs/test_mds_metrics.py:538]: * I think only 2 filesystems are required for successfully testing this issue. Why creating `fs_b` here?

File: qa/tasks/cephfs/test_mds_metrics.py [hunk 2]
```diff
@@ -511,3 +511,60 @@ def test_client_metrics_and_metadata(self):
...
+        # creating another filesystem
+        fs_c = self._setup_fs(fs_name="fs3")
+
+        # Mount a client on fs_c
+        self.mount_b.mount_wait(cephfs_name=fs_c.name)
+        self.mount_b.write_n_mb("test.bin", 1)
+        self.mount_b.path_to_ino("test.bin")
+        self.mount_b.create_files()
+
+        client_metadata = metrics['client_metadata']
```
Comment (joscollin) [inline qa/tasks/cephfs/test_mds_metrics.py:562]: you need call `self._get_metrics()` to get new set of metrics here.

File: qa/tasks/cephfs/test_mds_metrics.py [hunk 3]
```diff
@@ -511,3 +511,60 @@ def test_client_metrics_and_metadata(self):
...
+        global_metrics = metrics['global_metrics']
+
+        client_c_name = f'client.{self.mount_b.get_global_id()}'
+
+        if client_c_name not in client_metadata\
+                or client_c_name not in global_metrics:
+            raise RuntimeError("Client of new filesystem is missing!")
```
Comment (joscollin) [inline qa/tasks/cephfs/test_mds_metrics.py:569]: Good check. But it could be more specific like checking inside `client_metadata[fs_c.name]` ?

Also update the error message like `Invalid metrics after the filesystem {fs_c.name} fail`.

File: src/pybind/mgr/stats/fs/perf_stats.py [hunk 4]
```diff
@@ -195,15 +195,17 @@ def notify_fsmap(self):
...
-                #the new rank0 MDS is up:active
-                ua_last_updated = time.monotonic()
-                if (self.rqtimer and self.rqtimer.is_alive()):
-                    self.rqtimer.cancel()
-                self.rqtimer = Timer(REREGISTER_TIMER_INTERVAL,
-                                     self.re_register_queries, args=(rank0_gid, ua_last_updated,))
-                self.rqtimer.start()
+            for value in gid_state:
+                rank0_gid = value[0]
+                state = value[1]
```
Comment (joscollin) [inline src/pybind/mgr/stats/fs/perf_stats.py:200]: nit: could be a single statement as done in earlier code :)

File: src/pybind/mgr/stats/fs/perf_stats.py [hunk 5]
```diff
@@ -195,15 +195,17 @@ def notify_fsmap(self):
...
+                if (rank0_gid and rank0_gid != self.prev_rank0_gid and state == 'up:active'):
+                    #the new rank0 MDS is up:active
+                    ua_last_updated = time.monotonic()
+                    if (self.rqtimer and self.rqtimer.is_alive()):
+                        self.rqtimer.cancel()
+                    self.rqtimer = Timer(REREGISTER_TIMER_INTERVAL,
+                                        self.re_register_queries, args=(rank0_gid, ua_last_updated,))
```
Comment (joscollin) [inline src/pybind/mgr/stats/fs/perf_stats.py:207]: align here.

**Review by joscollin — CHANGES_REQUESTED** [review]

File: qa/tasks/cephfs/test_mds_metrics.py (hunk 1 above)
Comment (neesingh-rh) [reply qa/tasks/cephfs/test_mds_metrics.py:538]: Yeah, for that we can call `self._get_metrics()` initially to enable `stats module`.

File: qa/tasks/cephfs/test_mds_metrics.py (hunk 3 above)
Comment (neesingh-rh) [reply qa/tasks/cephfs/test_mds_metrics.py:569]: Actually, addition of fs_name in `perf stats` is not yet merged in main.So,we have to write it as per our earlier output structure until new changes comes in mainline.

Comment (joscollin) [reply qa/tasks/cephfs/test_mds_metrics.py:569]: That's okay, you can make the changes based on the new `perf stats` output. For local testing, add the other commit in this branch.

Comment (neesingh-rh) [reply qa/tasks/cephfs/test_mds_metrics.py]: If that's the case, I have added `mul_fs` field also.

File: qa/tasks/cephfs/test_mds_metrics.py [hunk 6]
```diff
@@ -511,3 +511,83 @@ def test_client_metrics_and_metadata(self):
...
+            log.debug(f'metrics={metrics_new}')
+            self.assertTrue(valid)
+
+            global_metrics = metrics_new['global_metrics']
+            client_a_metrics_new = global_metrics["fs1"][client_a_name]
+            client_b_metrics_new = global_metrics["fs2"][client_b_name]
+
+            # the metrics should be different for the test to succeed.
+            self.assertNotEqual(client_a_metrics, client_a_metrics_new)
+            self.assertNotEqual(client_b_metrics, client_b_metrics_new)
```
Comment (joscollin) [inline qa/tasks/cephfs/test_mds_metrics.py:587]: Lets check the `client_metadata` too, as they are also stale after the failover.

File: qa/tasks/cephfs/test_mds_metrics.py [hunk 7]
```diff
@@ -511,3 +511,83 @@ def test_client_metrics_and_metadata(self):
...
+
+        # creating filesystem
+        fs_a = self._setup_fs(fs_name="fs1")
+
+        # Mount a client on fs_a
+        self.mount_a.mount_wait(cephfs_name=fs_a.name)
+        self.mount_a.write_n_mb("pad.bin", 1)
+        self.mount_a.write_n_mb("test.bin", 2)
+        self.mount_a.path_to_ino("test.bin")
+        self.mount_a.create_files()
```
[the remark this answers is not in the document]
Comment (joscollin) [reply qa/tasks/cephfs/test_mds_metrics.py:534]: As discussed, `_spread_directory_on_all_ranks` is written specifically for  `mount_a`. So let's skip this change.

File: qa/tasks/cephfs/test_mds_metrics.py [hunk 8]
```diff
@@ -511,3 +516,75 @@ def test_client_metrics_and_metadata(self):
...
+            valid, metrics_new = self._get_metrics(
+                self.verify_mds_metrics(client_count=1, mul_fs=[fs_a.id, fs_b.id]), 30)
+            log.debug(f'metrics={metrics_new}')
+            self.assertTrue(valid)
+
+            global_metrics = metrics_new['global_metrics']
+            client_b_metrics_new = global_metrics["fs2"][client_b_name]
+
+            client_metadata = metrics_new['client_metadata']
+            client_b_metadata_new = client_metadata["fs2"]
```
Comment (joscollin) [inline qa/tasks/cephfs/test_mds_metrics.py:580]: oops, you'll get a keyerror here if the `global_metrics` and `client_metadata` are empty after the failover (mostly, it's empty). See how I fixed this issue here: https://github.com/ceph/ceph/pull/47166/files and finally a single assert did the check with an error message :)

File: qa/tasks/cephfs/test_mds_metrics.py [hunk 9]
```diff
@@ -31,23 +31,28 @@ def _start_with_single_active_mds(self):
...
-            global_metrics = metrics['global_metrics']
-            client_metadata = metrics['client_metadata']
-            if not len(global_metrics) >= client_count or not len(client_metadata) >= client_count:
-                return False
+            for item in mul_fs:
+                key = fs_status.get_fsmap(item)['mdsmap']['fs_name']
+                global_metrics = metrics['global_metrics'][key]
+                client_metadata = metrics['client_metadata'][key]
+                if not len(global_metrics) >= client_count or not len(client_metadata) >= client_count:
+                    return False
```
Comment (joscollin) [inline qa/tasks/cephfs/test_mds_metrics.py:55]: was this change intended to merge via another PR?

File: qa/tasks/cephfs/test_mds_metrics.py [hunk 10]
```diff
@@ -511,3 +516,75 @@ def test_client_metrics_and_metadata(self):
...
+
+            global_metrics = metrics_new['global_metrics']
+            client_b_metrics_new = global_metrics["fs2"][client_b_name]
+
+            client_metadata = metrics_new['client_metadata']
+            client_b_metadata_new = client_metadata["fs2"]
+
+            # the metrics should be different for the test to succeed.
+            self.assertNotEqual(client_b_metrics, client_b_metrics_new)
+            self.assertIn(client_b_name, client_b_metadata_new)
```
Comment (joscollin) [inline qa/tasks/cephfs/test_mds_metrics.py:584]: Let's do a single assert with the error message as in: https://github.com/ceph/ceph/pull/47166/files

File: qa/tasks/cephfs/test_mds_metrics.py [hunk 11]
```diff
@@ -511,3 +516,75 @@ def test_client_metrics_and_metadata(self):
...
+        global_metrics = metrics['global_metrics']
+        client_b_metrics = global_metrics["fs2"][client_b_name]
+
+        # fail active mds of fs_b
+        fs_b_mds = fs_b.get_active_names()[0]
+        self.mds_cluster.mds_fail(fs_b_mds)
+        fs_b.wait_for_daemons()
+
+        # spread directory per rank
+        self._spread_directory_on_all_ranks(fs_b.id)
```
Comment (joscollin) [inline qa/tasks/cephfs/test_mds_metrics.py:564]: Does this operate/write on `mount_a`? But looks like you've mounted `fs_b` on `mount_b`.

File: qa/tasks/cephfs/test_mds_metrics.py [hunk 12]
```diff
@@ -511,3 +516,75 @@ def test_client_metrics_and_metadata(self):
...
+
+        # Mount a client on fs_b
+        self.mount_b.mount_wait(cephfs_name=fs_b.name)
+        self.mount_b.write_n_mb("test.bin", 1)
+        self.mount_b.path_to_ino("test.bin")
+        self.mount_b.create_files()
+
+        # validate
+        valid, metrics = self._get_metrics(
+            self.verify_mds_metrics(client_count=1, mul_fs=[fs_a.id, fs_b.id]), 30)
```
Comment (joscollin) [inline qa/tasks/cephfs/test_mds_metrics.py:547]: Is `client_count=1` correct when `mul_fs` has two fs ids? Or it should be 2?
(I haven't checked your `mul_fs` patch).

**Review by joscollin — CHANGES_REQUESTED** [review]

File: qa/tasks/cephfs/test_mds_metrics.py [hunk 13]
```diff
@@ -511,3 +516,75 @@ def test_client_metrics_and_metadata(self):
...
+        # entries from the global_metrics.
+        client_b_name = f'client.{self.mount_b.get_global_id()}'
+
+        global_metrics = metrics['global_metrics']
+        client_b_metrics = global_metrics["fs2"][client_b_name]
+
+        # fail active mds of fs_b
+        fs_b_mds = fs_b.get_active_names()[0]
+        self.mds_cluster.mds_fail(fs_b_mds)
+        fs_b.wait_for_daemons()
```
Comment (joscollin) [inline qa/tasks/cephfs/test_mds_metrics.py:561]: Sorry, I was mistaken last time. As I checked the code again, it looks like `self.fs_b.wait_for_state('up:active', rank=0, timeout=30)` would be more appropriate here as it operates on `fs_b`. So let's wait for `fs_b`'s rank0 mds to be active state instead. That's better and assures getting the metrics.

**Review by joscollin — CHANGES_REQUESTED** [review]

File: qa/tasks/cephfs/test_mds_metrics.py (hunk 12 above)
Comment (neesingh-rh) [reply qa/tasks/cephfs/test_mds_metrics.py:542]: Yup, check is done for each fs separately.

File: qa/tasks/cephfs/test_mds_metrics.py [hunk 14]
```diff
@@ -511,3 +511,76 @@ def test_client_metrics_and_metadata(self):
...
+        # mount_a is the client mounted for TestMDSMetrics. So get its
+        # entry from the global_metrics.
+        client_a_name = f'client.{self.mount_a.get_global_id()}'
+
+        global_metrics = metrics['global_metrics']
+        client_a_metrics = global_metrics.get("fs1", {}).get(client_a_name, {})
+
+        # fail active mds of fs_a
+        fs_a_mds = fs_a.get_active_names()[0]
+        self.mds_cluster.mds_fail(fs_a_mds)
```
Comment (joscollin) [inline qa/tasks/cephfs/test_mds_metrics.py:555]: Now this is a wrong test. Because `perf stats` is already resilient to the first created fs's mds fail. You need to fail the second created one (the second one in fsmap) or the third one instead to reproduce the stale/empty metrics issue. Just rearrange the creation order above.

**Review by joscollin — CHANGES_REQUESTED** [review]

**Review by joscollin — CHANGES_REQUESTED** [review]
otherwise, looks good to me!

File: qa/tasks/cephfs/test_mds_metrics.py [hunk 15]
```diff
@@ -511,3 +511,77 @@ def test_client_metrics_and_metadata(self):
...
+        self.mount_a.path_to_ino("test.bin")
+        self.mount_a.create_files()
+
+        # validate
+        valid, metrics = self._get_metrics(
+            self.verify_mds_metrics(client_count=1, mul_fs=[fs_a.id, fs_b.id]), 30)
+        log.debug(f"metrics={metrics}")
+        self.assertTrue(valid)
+
+        # mount_a is the client mounted for TestMDSMetrics. So get its
```
Comment (joscollin) [inline qa/tasks/cephfs/test_mds_metrics.py:546]: This comment doesn't make sense here, as we unmounted them initially and mounted our own fs.

**Review by joscollin — APPROVED** [review]

File: qa/tasks/cephfs/test_mds_metrics.py [hunk 16]
```diff
@@ -511,3 +511,78 @@ def test_client_metrics_and_metadata(self):
...
+        self.mds_cluster.mds_fail(fs_a_mds)
+        fs_a.wait_for_state('up:active', rank=0, timeout=30)
+
+        # spread directory per rank
+        self._spread_directory_on_all_ranks(fs_a.id)
+
+        # spread some I/O
+        self._do_spread_io_all_clients(fs_a.id)
+
+        time.sleep(5)
```
Comment (joscollin) [inline qa/tasks/cephfs/test_mds_metrics.py:563]: Please retain the comment too, because it's important to know why we are waiting.

**Review by joscollin — CHANGES_REQUESTED** [review]

**Review by joscollin — APPROVED** [review]

**Comment by vshankar** [comment]
dropping these changes from test branch - @joscollin @neesingh-rh please reach out once the issues have been fixed.

**Comment by neesingh-rh** [comment]
jenkins test windows

**Comment by joscollin** [comment]
Test result:
http://pulpito.front.sepia.ceph.com/jcollin-2022-08-17_13:13:50-fs-wip-jcollin-perf-stats-fstop-testing-distro-default-smithi/

@neesingh-rh 
`test_perf_stats_stale_metrics_with_multiple_filesystem` fails with `KeyError: 'fs1'`. Note that the similar test `test_perf_stats_stale_metrics` succeeds. Please check.

**Comment by joscollin** [comment]
Tests passed, the failures are unrelated:
http://pulpito.front.sepia.ceph.com/jcollin-2022-08-22_06:28:29-fs-wip-jcollin-perf-stats-fstop-testing_18_Aug_22-distro-default-smithi/

@vshankar Please take a look and merge.

**Review by vshankar — APPROVED** [review]
